src/objective/regression_obj.cu

```diff
@@ -73,29 +73,43 @@ class RegLossObj : public ObjFunction {
     additional_input_.HostVector().begin()[1] = scale_pos_weight;
     additional_input_.HostVector().begin()[2] = is_null_weight;
 
-    common::Transform<>::Init([] XGBOOST_DEVICE(size_t _idx,
+    const size_t nthreads = omp_get_max_threads();
+    const size_t n_data_blocks = nthreads;
+    common::Transform<>::Init([ndata, nthreads] XGBOOST_DEVICE(size_t data_block_idx,
                            common::Span<float> _additional_input,
                            common::Span<GradientPair> _out_gpair,
                            common::Span<const bst_float> _preds,
                            common::Span<const bst_float> _labels,
                            common::Span<const bst_float> _weights) {
+          const bst_float* preds_ptr = _preds.data();
+          const bst_float* labels_ptr = _labels.data();
+          const bst_float* weights_ptr = _weights.data();
+          GradientPair* out_gpair_ptr = _out_gpair.data();
+          const size_t block_size = ndata/nthreads + !!(ndata%nthreads);
```

Comment:
Is this optimization really necessary?  The objective is only run once for each iteration.  Also, have you thought about extracting the "CheckLabel" out as an independent function?  We can run it once at the beginning of training.  Removing 1 condition seems helpful?

Reply:
Yes. The most of the hotspots here is due to spans (`_additional_input`, `out_gpair`,  `_preds`,  `_labels`, `_weights`) are created for each individual samples, and extra checks in span `operator[]`. So here I tried to minimize span creation by reducing number of blocks.